### ISSUE TITLE
Add support for configurable prompt pattern

### ISSUE THREAD

**Issue by saparikh** [jeremyschulman/netcfgbu#86]
The built-in prompt pattern doesn't work for Cumulus devices. Because of this `pre_get_config` is not usable.

You end up with this error when trying to use `pre_get_config`:

```
FAILURES: 20
host     os_name    reason
-------  ---------  -----------------------------------
leaf02   cumulus    TIMEOUT('Timeout awaiting prompt',)
leaf03   cumulus    TIMEOUT('Timeout awaiting prompt',)
leaf13   cumulus    TIMEOUT('Timeout awaiting prompt',)
```

The need for `pre_get_config` is coming from the fact that using this setting `get_config = "( cat /etc/hostname; cat /etc/network/interfaces; cat /etc/cumulus/ports.conf; sudo cat /etc/frr/frr.conf)"` results in the first few lines still being lost.

This is what should be retrieved:
```
cumulus@leaf01:mgmt-vrf:~$ ( cat /etc/hostname; cat /etc/network/interfaces; cat /etc/cumulus/ports.conf; sudo cat /etc/frr/frr.conf )
leaf01
# This file describes the network interfaces available on your system
# and how to activate them. For more information, see interfaces(5).

source /etc/network/interfaces.d/*.intf

auto lo
iface lo inet loopback
    address 10.16.1.1/32
    address 192.168.123.1/32
```

This is what is currently retrieved:
```
rmation, see interfaces(5).

source /etc/network/interfaces.d/*.intf

auto lo
iface lo inet loopback
    address 10.16.1.1/32
    address 192.168.123.1/32
```

**Comment by jeremyschulman** [comment]
I will add the ability to define a `prompt` pattern in the `[os_name.$name]` section.

**Comment by saparikh** [comment]
awesome.

I think this `:mgmt-vrf:~$` in the prompt is what is causing the existing regex to not work. I can play around with regexes to make this work.

Cisco IOS doesn't show the username in the prompt, it just looks like `exitgw#`. 
Juniper does, like this `juniper@fwl01>` which is why it works with the regex.

**Comment by jeremyschulman** [comment]
working on this today @saparikh

**Comment by jeremyschulman** [comment]
Confirming that the sample prompt does not match the default basic PROMPT_PATTERN:
```python
In [13]: res = PROMPT_PATTERN.match("cumulus@leaf01:mgmt-vrf:~$".encode('utf-8'))

In [14]: res is None
Out[14]: True
````

**Comment by jeremyschulman** [comment]
The tilde (~) is not present in the definition of `consts.PROMPT_VALID_CHARS`.

```python
PROMPT_VALID_CHARS = r"a-z0-9.\-_@()/:"
```

**Comment by jeremyschulman** [comment]
@saparikh created branch with enhancement:
https://github.com/jeremyschulman/netcfgbu/tree/jls-devAug09

Could you please dev-test in your enviornment?

**Comment by saparikh** [comment]
Awesome, thanks. I will give it a go later today.

**Comment by saparikh** [comment]
I am running into this issue at the moment:

```
(cfgbu) samir-mbp15:cicd_book saparikh$ netcfgbu backup --limit host=leaf01
2020-08-09 14:36:39,947 INFO: JUMPHOST: connected to samir@cosmos4.intentionet.com
2020-08-09 14:36:39,948 INFO: LOGIN: leaf01 (cumulus) timeout=60s as cumulus
2020-08-09 14:36:50,264 INFO: CONNECTED: leaf01
2020-08-09 14:36:50,487 INFO: CLOSED: leaf01
2020-08-09 14:36:50,488 INFO: DONE (1/1): leaf01 FALSE
# ------------------------------------------------------------------------------
Summary: TOTAL=1, OK=0, FAIL=1
         START=2020-Aug-09 02:36:39 PM, STOP=2020-Aug-09 02:36:50 PM
         DURATION=10.890s


FAILURES: 1
host    os_name    reason
------  ---------  -------------------------
leaf01  cumulus    IndexError: no such group
```

I have installed `netcfgbu` from the branch

```
(cfgbu) samir-mbp15:cicd_book saparikh$ pip freeze
aiofiles==0.5.0
asyncssh==2.2.1
cffi==1.14.1
click==7.1.2
cryptography==3.0
first==2.0.2
invoke==1.4.1
-e git+git@github.com:jeremyschulman/netcfgbu.git@905db5464e21ee7fa8b3c54c5faaa416b91820c9#egg=netcfgbu
pexpect==4.8.0
ptyprocess==0.6.0
pycparser==2.20
pydantic==1.5.1
six==1.15.0
tabulate==0.8.7
toml==0.10.1
```

**Comment by saparikh** [comment]
This is the portion of my toml file:
```
[os_name.cumulus]
    # NOTE: make sure that the user has password-less sudo access, otherwise the
    # get_config execution will fail. There is no current workaround for this
    # requirement. Also pre_get_config does not work for Cumulus devices at this time.
    #
    # Do not change the order of the cat commands either. This ensures the final
    # file format is recognized correctly by Batfish (https://github.com/batfish/batfish)

    pre_get_config = [
        "( echo $HOSTNAME; cat -u /etc/hostname; cat -u /etc/network/interfaces; cat -u /etc/cumulus/ports.conf; sudo cat -u /etc/frr/frr.conf) > running.cfg"
    ]

    get_config = "cat running.cfg"

    # example Cumulus prompt value: cumulus@leaf01:mgmt-vrf:~$
    prompt_pattern = '[a-z0-9.\-@:~]{10,65}\s*[#$]'
```

will try this without using `pre_get_config`

**Comment by saparikh** [comment]
Ok, so if I don't use `pre_get_config` I get the file, but has the same issue as before where the first few lines are cut out.

**Comment by jeremyschulman** [comment]
issue identified and I will resolve in the next commit.  RCA showed that the driver code requires the use of a regex capture group.  I need to update the code handling the User defined `prompt_pattern` to add the capture group on behalf of the User.

**Comment by jeremyschulman** [comment]
@saparikh - just pushed update to resolve the capture group issue.  Please re-test when you can.  I also made another tweak to the regex handling to match the default behavior.

**Comment by jeremyschulman** [comment]
Received confirmation from @saparikh that the latest code works AOK.  Going to merge this into master and release v0.6.0.

**Comment by saparikh** [comment]
Issue resolved in v0.6.0
